File: tests/data/program_analysis/SIR-Gillespie-SD_inline.f
C     Fortranification of AMIDOL's SIR-Gillespie.py

********************************************************************************

********************************************************************************
C     Variables:
C     beta     Rate of infection
C     gamma    Rate of recovery from an infection
C     rho      Basic reproduction Number
C
C     State Variables: S, I, R
C     S - Susceptible population
C     I - Infected population
C     R - Recovered population
C     n_S      current number of susceptible members
C     n_I      current number of infected members
C     n_R      current number of recovered members
C     S0       initial value of S
C     I0       initial value of I
C     R0       initial value of R
C     MeanS    Measures of Mean for S
C     MeanI    Measures of Mean for I
C     MeanR    Measures of Mean for R
C     VarS     Measures of Variance for S
C     VarI     Measures of Variance for I
C     VarR     Measures of Variance for R
C
C     rateInfect    Current state dependent rate of infection
C     rateRecover   Current state dependent rate of recovery
C     totalRates    Sum of total rates; taking advantage of Markovian identities
C                       to improve performance.
C
C     Tmax          Maximum time for the simulation
C     t             Initial time for the simulation
C     totalRuns     Total number of trajectories to generate for the analysis
C     dt       next inter-event time
********************************************************************************
      subroutine gillespie(S, I, R, gamma, rho)
        integer S, I, R
        integer, parameter :: Tmax = 100, total_runs = 1000
        double precision gamma, rho
        double precision, parameter :: beta = rho * gamma !
        double precision, dimension(0:Tmax) :: MeanS, MeanI, MeanR
        double precision, dimension(0:Tmax) :: VarS, VarI, VarR
        integer, dimension(0:Tmax) :: samples

        integer j, runs, n_S, n_I, n_R, sample_idx, samp, runs1
        double precision totalRates, dt, t, randval
        double precision rateInfect, rateRecover

        do j = 0, Tmax    ! Initialize the mean and variance arrays
          MeanS(j) = 0
          MeanI(j) = 0.0
          MeanR(j) = 0.0

          VarS(j) = 0.0
          VarI(j) = 0.0
          VarR(j) = 0.0

          samples(j) = j
        end do

        do runs = 0, total_runs-1
          t = 0.0    ! Restart the event clock

          ! main Gillespie loop
          sample_idx = 0
          do while (t .le. Tmax .and. I .gt. 0)
            n_S = S
            n_I = I
            n_R = R

            rateInfect = beta * S * I / (S + I + R)
            rateRecover = gamma * I
            totalRates = rateInfect + rateRecover

            ! Determine which event fired.  With probability rateInfect/totalRates
            ! the next event is infection.
            if (rand() < (rateInfect/totalRates)) then
              ! Delta for infection
              S = S - 1
              I = I + 1
            ! Determine the event fired.  With probability rateRecover/totalRates
            ! the next event is recovery.
            else
              ! Delta for recovery
              I = I - 1
              R = R + 1
            endif

            dt = -log(1.0-rand())/totalRates  ! next inter-event time
            t = t + dt          !  Advance the system clock

            ! Calculate all measures up to the current time t using
            ! Welford's one pass algorithm
            do while (sample_idx < Tmax .and. t > samples(sample_idx))
              samp = samples(sample_idx)

              runs1 = runs+1
              MeanS(samp) = MeanS(samp)+(n_S-MeanS(samp))/(runs1)
              VarS(samp) = VarS(samp) + runs/(runs1) *
     &                        (n_S-MeanS(samp))*(n_S-MeanS(samp))

              MeanI(samp) = MeanI(samp)+(n_I-MeanI(samp))/(runs1)
              VarI(samp) = VarI(samp) + runs/(runs1) *
     &                         (n_I-MeanI(samp))*(n_I-MeanI(samp))

              MeanR(samp) = MeanR(samp) + (n_R - MeanR(samp))/(runs1)
              VarR(samp) = VarR(samp) + runs/(runs1) *
     &                         (n_R-MeanR(samp))*(n_R-MeanR(samp))

              sample_idx = sample_idx+1
            end do
          end do

          ! After all events have been processed, clean up by evaluating all remaining measures.
          do while (sample_idx < Tmax)
            samp = samples(sample_idx)

            runs1 = runs+1
            MeanS(samp) = MeanS(samp)+(n_S-MeanS(samp))/(runs1)
            VarS(samp) = VarS(samp) + runs/(runs1) *
     &                        (n_S-MeanS(samp))*(n_S-MeanS(samp))

            MeanI(samp) = MeanI(samp)+(n_I-MeanI(samp))/(runs1)
            VarI(samp) = VarI(samp) + runs/(runs1) *
     &                         (n_I-MeanI(samp))*(n_I-MeanI(samp))

            MeanR(samp) = MeanR(samp) + (n_R - MeanR(samp))/(runs1)
            VarR(samp) = VarR(samp) + runs/(runs1) *
     &                         (n_R-MeanR(samp))*(n_R-MeanR(samp))

            sample_idx = sample_idx + 1
          end do
        end do
      end subroutine gillespie

      program main
      integer, parameter :: S = 500, I = 10, R = 0, Tmax = 100
      double precision, parameter :: gamma = 1.0/3.0, rho = 2.0

      call gillespie(S, I, R, gamma, rho)
      end program main

File: tests/data/program_analysis/SIR-Gillespie-SD_inline.py
import sys
import os
from typing import List
import math
from delphi.translators.for2py.format import *
from delphi.translators.for2py.arrays import *
from delphi.translators.for2py.static_save import *
from delphi.translators.for2py.strings import *
from delphi.translators.for2py import intrinsics
from dataclasses import dataclass
from delphi.translators.for2py.types_ext import Float32
import delphi.translators.for2py.math_ext as math
from numbers import Real
from random import random


def gillespie(s: List[int], i: List[int], r: List[int], gamma: List[float], rho: List[float]):
    tmax: List[int] = [100]
    total_runs: List[int] = [1000]
    beta: List[float] = [(rho[0] * gamma[0])]
    means = Array(float, [(0, tmax[0])])
    meani = Array(float, [(0, tmax[0])])
    meanr = Array(float, [(0, tmax[0])])
    vars = Array(float, [(0, tmax[0])])
    vari = Array(float, [(0, tmax[0])])
    varr = Array(float, [(0, tmax[0])])
    samples = Array(int, [(0, tmax[0])])
    j: List[int] = [None]
    runs: List[int] = [None]
    n_s: List[int] = [None]
    n_i: List[int] = [None]
    n_r: List[int] = [None]
    sample_idx: List[int] = [None]
    samp: List[int] = [None]
    runs1: List[int] = [None]
    totalrates: List[float] = [None]
    dt: List[float] = [None]
    t: List[float] = [None]
    randval: List[float] = [None]
    rateinfect: List[float] = [None]
    raterecover: List[float] = [None]
    for j[0] in range(0, tmax[0]+1):
        means.set_((j[0]), Float32(0))
        meani.set_((j[0]), Float32(0.0))
        meanr.set_((j[0]), Float32(0.0))
        vars.set_((j[0]), Float32(0.0))
        vari.set_((j[0]), Float32(0.0))
        varr.set_((j[0]), Float32(0.0))
        samples.set_((j[0]), int(j[0]))
    for runs[0] in range(0, (total_runs[0] - 1)+1):
        t[0] = 0.0
        sample_idx[0] = 0
        while ((t[0] <= tmax[0]) and (i[0] > 0)):
            n_s[0] = s[0]
            n_i[0] = i[0]
            n_r[0] = r[0]
            rateinfect[0] = (((beta[0] * s[0]) * i[0]) / ((s[0] + i[0]) + r[0]))
            raterecover[0] = (gamma[0] * i[0])
            totalrates[0] = (rateinfect[0] + raterecover[0])
            if (random() < (rateinfect[0] / totalrates[0])):
                s[0] = (s[0] - 1)
                i[0] = (i[0] + 1)
            else:
                i[0] = (i[0] - 1)
                r[0] = (r[0] + 1)
            dt[0] = -((math.log((1.0 - random())) / totalrates[0]))
            t[0] = (t[0] + dt[0])
            while ((sample_idx[0] < tmax[0]) and (t[0] > samples.get_((sample_idx[0])))):
                samp[0] = samples.get_((sample_idx[0]))
                runs1[0] = (runs[0] + 1)
                means.set_((samp[0]), Float32((means.get_((samp[0])) + ((n_s[0] - means.get_((samp[0]))) / runs1[0]))))
                vars.set_((samp[0]), Float32((vars.get_((samp[0])) + (((runs[0] / runs1[0]) * (n_s[0] - means.get_((samp[0])))) * (n_s[0] - means.get_((samp[0])))))))
                meani.set_((samp[0]), Float32((meani.get_((samp[0])) + ((n_i[0] - meani.get_((samp[0]))) / runs1[0]))))
                vari.set_((samp[0]), Float32((vari.get_((samp[0])) + (((runs[0] / runs1[0]) * (n_i[0] - meani.get_((samp[0])))) * (n_i[0] - meani.get_((samp[0])))))))
                meanr.set_((samp[0]), Float32((meanr.get_((samp[0])) + ((n_r[0] - meanr.get_((samp[0]))) / runs1[0]))))
                varr.set_((samp[0]), Float32((varr.get_((samp[0])) + (((runs[0] / runs1[0]) * (n_r[0] - meanr.get_((samp[0])))) * (n_r[0] - meanr.get_((samp[0])))))))
                sample_idx[0] = (sample_idx[0] + 1)
        while (sample_idx[0] < tmax[0]):
            samp[0] = samples.get_((sample_idx[0]))
            runs1[0] = (runs[0] + 1)
            means.set_((samp[0]), Float32((means.get_((samp[0])) + ((n_s[0] - means.get_((samp[0]))) / runs1[0]))))
            vars.set_((samp[0]), Float32((vars.get_((samp[0])) + (((runs[0] / runs1[0]) * (n_s[0] - means.get_((samp[0])))) * (n_s[0] - means.get_((samp[0])))))))
            meani.set_((samp[0]), Float32((meani.get_((samp[0])) + ((n_i[0] - meani.get_((samp[0]))) / runs1[0]))))
            vari.set_((samp[0]), Float32((vari.get_((samp[0])) + (((runs[0] / runs1[0]) * (n_i[0] - meani.get_((samp[0])))) * (n_i[0] - meani.get_((samp[0])))))))
            meanr.set_((samp[0]), Float32((meanr.get_((samp[0])) + ((n_r[0] - meanr.get_((samp[0]))) / runs1[0]))))
            varr.set_((samp[0]), Float32((varr.get_((samp[0])) + (((runs[0] / runs1[0]) * (n_r[0] - meanr.get_((samp[0])))) * (n_r[0] - meanr.get_((samp[0])))))))
            sample_idx[0] = (sample_idx[0] + 1)

def main():
    s: List[int] = [500]
    i: List[int] = [10]
    r: List[int] = [0]
    tmax: List[int] = [100]
    gamma: List[float] = [(1.0 / 3.0)]
    rho: List[float] = [2.0]
    gillespie(s, i, r, gamma, rho)

main()
